pkg/types/timestamp/v0.0.1/entry.go

```diff
@@ -0,0 +1,179 @@
+	"errors"
+	"fmt"
+
+	"github.com/asaskevich/govalidator"
+	"github.com/go-openapi/strfmt"
+	"github.com/go-openapi/swag"
+	"github.com/sigstore/rekor/pkg/types/timestamp"
+
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"github.com/sigstore/rekor/pkg/log"
+	"github.com/sigstore/rekor/pkg/types"
+)
+
+const (
+	APIVERSION = "0.0.1"
+)
+
+func init() {
+	if err := timestamp.VersionMap.SetEntryFactory(APIVERSION, NewEntry); err != nil {
+		log.Logger.Panic(err)
+	}
+}
+
+type V001Entry struct {
+	TimestampObj models.TimestampV001Schema
+}
+
+func (v V001Entry) APIVersion() string {
+	return APIVERSION
+}
+
+func NewEntry() types.EntryImpl {
+	return &V001Entry{}
+}
+
+func (v V001Entry) IndexKeys() []string {
+	var result []string
+
+	if v.HasExternalEntities() {
+		if err := v.FetchExternalEntities(context.Background()); err != nil {
+			log.Logger.Error(err)
+			return result
+		}
+	}
+
+	if sha, err := contentToSha256(v.TimestampObj.Tsr.Content); err == nil {
+		payloadKey := "sha256:" + sha
+		result = append(result, payloadKey)
+	} // should we log the error here?
+
+	return result
+}
+
+func (v *V001Entry) Unmarshal(pe models.ProposedEntry) error {
+	timestampResp, ok := pe.(*models.Timestamp)
+	if !ok {
+		return errors.New("cannot unmarshal non Timestamp v0.0.1 type")
+	}
+
+	if err := types.DecodeEntry(timestampResp.Spec, &v.TimestampObj); err != nil {
+		return err
+	}
+
+	// field validation
+	if err := v.TimestampObj.Validate(strfmt.Default); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (v V001Entry) HasExternalEntities() bool {
+	return false
+}
+
+func (v *V001Entry) FetchExternalEntities(ctx context.Context) error {
+	return nil
+}
+
+func (v *V001Entry) Canonicalize(ctx context.Context) ([]byte, error) {
+	if err := v.FetchExternalEntities(ctx); err != nil {
+		return nil, err
+	}
+
+	if v.TimestampObj.Tsr == nil || v.TimestampObj.Tsr.Content == nil {
+		return nil, fmt.Errorf("tsr content is nil")
+	}
+	// compute the sha256 hash of the tsr content
+	sha, err := contentToSha256(v.TimestampObj.Tsr.Content)
+	if err != nil {
+		return nil, err
+	}
+
+	content := *v.TimestampObj.Tsr.Content
+
+	canonicalEntry := models.TimestampV001Schema{
+		Tsr: &models.TimestampV001SchemaTsr{
+			Content: &content,
+			Hash: &models.TimestampV001SchemaTsrHash{
+				Algorithm: swag.String(models.TimestampV001SchemaTsrHashAlgorithmSha256),
+				Value:     swag.String(sha),
+			},
+		},
+	}
+
+	// ExtraData is copied through unfiltered
+	canonicalEntry.ExtraData = v.TimestampObj.ExtraData
+
+	// wrap in valid object with kind and apiVersion set
+	timestampObj := models.Timestamp{}
+	timestampObj.APIVersion = swag.String(APIVERSION)
+	timestampObj.Spec = &canonicalEntry
+
+	bytes, err := json.Marshal(&timestampObj)
+	if err != nil {
+		return nil, err
+	}
+
+	return bytes, nil
+}
+
+// Validate performs cross-field validation for fields in object
+func (v V001Entry) Validate() error {
+
+	// TODO: validate the tsr content here?
```

Comment:
@asraa any idea what would make sense to validate here?

Reply:
Hi hi! So probably you want to validate the ContentInfoSignedData contained in the TimeStampResp. I would:
(1) asn1 unmarshal the response into a [pkcs9.TimeStampResp](https://pkg.go.dev/github.com/sassoftware/relic@v7.1.0+incompatible/lib/pkcs9#TimeStampResp) resp
(2) If the resp.Status.Status is not Granted/GrantedWithMods (0 or 1) then there should not be any TimeStampToken attached and maybe we allow people uploading failed but well-formed timestamps. So stop here
(3) Otherwise, we have a TimeStampToken (type [pkcs7.ContentInfoSignedData](https://pkg.go.dev/github.com/sassoftware/relic@v7.1.0+incompatible/lib/pkcs7#ContentInfoSignedData)). Verify that the ContentType is a signed data OID attribute `{1, 2, 840, 113549, 1, 7, 2}`. 
(4) Verify the pkcs7.SignedData attribute (TimeStampToken.Content) in the token using this [Verify](https://github.com/sassoftware/relic/blob/v7.1.0/lib/pkcs7/verify.go#L44) method, with no external content, but skipDigests false.
(5) Final note: we can't really verify the cert chain with a given root as suggested in the doc string from (4), so skip that. We've validated the sig/content.

Reply:
Thanks! This makes sense, for now though, I failed if status is not: Granted/GrantedWithMods